[WIP] cli: backtrace integration

petermattis opened the pull request:
Initialize backtrace if available.

Use `stopper.Recover` in the connection handling goroutines.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7980)

<!-- Reviewable:end -->

petermattis committed:
cli: backtrace integration

Initialize backtrace if available.

Use `stopper.Recover` in the connection handling goroutines.

petermattis commented:
Seems to work. A dump is generated when a panic is received which I can inspect using the `hydra` tool.

tamird commented on cli/backtrace.go:
```diff
@@ -0,0 +1,80 @@
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
```
Shouldn't this be on the stopper?

petermattis replied:
Yes. It should. Good catch.

tamird commented on sql/show.go:
```diff
@@ -48,6 +48,9 @@ func (p *planner) Show(n *parser.Show) (planNode, error) {
 	case `TRANSACTION PRIORITY`:
 		v.rows = append(v.rows, []parser.Datum{parser.NewDString(p.txn.UserPriority.String())})
+	case `PANIC`:
+		// TODO(peter): Remove. This is just for testing.
```
?

petermattis replied:
I'm going to remove this before merging this PR. I wanted to be able to force a panic during SQL execution in order to test the panic handling. This allowed me to create a panic by executing the statement `SHOW PANIC`.

tamird commented:
LGTM

petermattis commented:
Added support for dumping traces on signal reception and for hooking `log.Fatal*`. And removed the changed `builder.sh` because backtrace doesn't work in docker (currently). PTAL.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 9 of 9 files at r1, 6 of 6 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[cli/backtrace.go, line 73 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNHzNXdFDGOr6neaMgc:-KNHzNXdFDGOr6neaMgd:-81268349) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0b72abaf006a5b93f0c5e9a9324845f2bc5cdcc/cli/backtrace.go#L73)):*
> ```Go
> 	bcd.Register(tracer)
> 
> 	// Hook log.Fatal*.
> ```

Slightly off topic, but do we really need `log.Fatal` to do anything but panic?

---

*[cli/backtrace.go, line 88 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNHzZOk79NR_THkUI9-:-KNHzZOk79NR_THkUI90:746689460) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0b72abaf006a5b93f0c5e9a9324845f2bc5cdcc/cli/backtrace.go#L88)):*
> ```Go
> 	}))
> 
> 	log.Infof("backtrace enabled")
> ```

Log the output file?

---

*[cli/backtrace.go, line 97 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI-64SmwFo1AzFCtwt:-KNI-64SmwFo1AzFCtwu:367783074) ([raw file](https://github.com/cockroachdb/cockroach/blob/e847198c7f680c081cc376614f14441c018e0e49/cli/backtrace.go#L97)):*
> ```Go
> 	// tracer can only print to stdout/err).
> 	if f, err := os.Create(filepath.Join(logDir, "backtrace.out")); err != nil {
> 		log.Infof("%s", err)
> ```

Add a message here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-:-KNI-DkQUEQ_HqgVHFH0:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/backtrace.go, line 95 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI-QaN4hbBDB4Y464w:-KNI-QaN4hbBDB4Y464x:-1056372613) ([raw file](https://github.com/cockroachdb/cockroach/blob/e847198c7f680c081cc376614f14441c018e0e49/cli/backtrace.go#L95)):*
> ```Go
> 	// Tracer I/O is directed to os.DevNull by default.
> 	// Note: this does not affect the generated output file (unless the
> 	// tracer can only print to stdout/err).
> ```

I don't understand this comment.

---

*[cli/backtrace.go, line 96 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI-S0_agMF2jJasbhW:-KNI-S0_agMF2jJasbhX:1208545732) ([raw file](https://github.com/cockroachdb/cockroach/blob/e847198c7f680c081cc376614f14441c018e0e49/cli/backtrace.go#L96)):*
> ```Go
> 	// Note: this does not affect the generated output file (unless the
> 	// tracer can only print to stdout/err).
> 	if f, err := os.Create(filepath.Join(logDir, "backtrace.out")); err != nil {
> ```

Is there any need to make that file unique? I'm thinking running three nodes locally at the same time, etc. Wouldn't want clobbering here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 10 of 11 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[cli/backtrace.go, line 73 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNHzNXdFDGOr6neaMgc:-KNI04VwV7IqOrdUiJmf:-987742444) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0b72abaf006a5b93f0c5e9a9324845f2bc5cdcc/cli/backtrace.go#L73)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Slightly off topic, but do we really need `log.Fatal` to do anything but panic?
</details>
Maybe. That's a task for another PR.

---

*[cli/backtrace.go, line 88 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNHzZOk79NR_THkUI9-:-KNI0Kbxa0Th-pxYKlQy:-1338455192) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0b72abaf006a5b93f0c5e9a9324845f2bc5cdcc/cli/backtrace.go#L88)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Log the output file?
</details>
Which output file? The`*.btt` files are written to the current directory, but we don't know the filename until they are generated.

---

*[cli/backtrace.go, line 95 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI-QaN4hbBDB4Y464w:-KNI08R1mN4OtcTDXj9c:1924478169) ([raw file](https://github.com/cockroachdb/cockroach/blob/e847198c7f680c081cc376614f14441c018e0e49/cli/backtrace.go#L95)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I don't understand this comment.
</details>
Copied&pasted from their example. I'll fix it up.

---

*[cli/backtrace.go, line 96 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI-S0_agMF2jJasbhW:-KNI0CU_HBV-YaH32SGI:-1644034766) ([raw file](https://github.com/cockroachdb/cockroach/blob/e847198c7f680c081cc376614f14441c018e0e49/cli/backtrace.go#L96)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Is there any need to make that file unique? I'm thinking running three nodes locally at the same time, etc. Wouldn't want clobbering here.
</details>
We're putting this file in the per-node log directory. No additional uniqueness required.

---

*[cli/backtrace.go, line 97 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI-64SmwFo1AzFCtwt:-KNI0HrmNZg2eCWwXKfi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e847198c7f680c081cc376614f14441c018e0e49/cli/backtrace.go#L97)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Add a message here.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r4.
Review status: 10 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[cli/backtrace.go, line 93 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI1xTzMGdKFHve966E:-KNI1xU-ShEQSXHV9hqJ:545322586) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f2750b47f4521aa532361f797a9ff4bf23d56fb/cli/backtrace.go#L93)):*
> ```Go
> 	}))
> 
> 	// Internal, backtrace uses an internal program (/opt/backtrace/bin/ptrace)
> ```

s/Internal/Internally/
s/internal/ptrace/

---

*[cli/backtrace.go, line 96 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI27O9XeI0qpQOKtas:-KNI27O9XeI0qpQOKtat:595906261) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f2750b47f4521aa532361f797a9ff4bf23d56fb/cli/backtrace.go#L96)):*
> ```Go
> 	// to generate traces. We direct the stdout for this program to a file for
> 	// debugging our usage of backtrace.
> 	if f, err := os.Create(filepath.Join(logDir, "backtrace.out")); err != nil {
> ```

Should this be opened in append mode or are you intentionally truncating every time the process starts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[cli/backtrace.go, line 93 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI1xTzMGdKFHve966E:-KNIhUD90ZbD3-2qRvnu:-483721578) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f2750b47f4521aa532361f797a9ff4bf23d56fb/cli/backtrace.go#L93)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/Internal/Internally/
> s/internal/ptrace/
</details>
Ugh, not my day for grammar. Fixed.

---

*[cli/backtrace.go, line 96 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNI27O9XeI0qpQOKtas:-KNIhWZRkWvTzq206Nnh:-1154112895) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f2750b47f4521aa532361f797a9ff4bf23d56fb/cli/backtrace.go#L96)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Should this be opened in append mode or are you intentionally truncating every time the process starts?
</details>
I'm not even sure if this file is useful. Switched to appending. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 9 files at r1, 5 of 6 files at r2, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[cli/backtrace.go, line 99 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNIiagjc_YXWWN7MNoc:-KNIiagjc_YXWWN7MNod:897867799) ([raw file](https://github.com/cockroachdb/cockroach/blob/094383335405d59c25ec8708e4e9da11dbc737ab/cli/backtrace.go#L99)):*
> ```Go
> 		log.Infof("unable to open: %s", err)
> 	} else {
> 		stopper.AddCloser(stop.CloserFn(func() {
> ```

can't you just `stopper.AddCloser(f)` here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cli/backtrace.go, line 99 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNIiagjc_YXWWN7MNoc:-KNIjw4xzAocYHZvrUHf:-1209258908) ([raw file](https://github.com/cockroachdb/cockroach/blob/094383335405d59c25ec8708e4e9da11dbc737ab/cli/backtrace.go#L99)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can't you just `stopper.AddCloser(f)` here?
</details>
`File.Close` returns an `error`. `stopper.Closer` does not. :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cli/backtrace.go, line 99 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980#-KNIiagjc_YXWWN7MNoc:-KNIkok_qrPTFkXLZA4R:-332565974) ([raw file](https://github.com/cockroachdb/cockroach/blob/094383335405d59c25ec8708e4e9da11dbc737ab/cli/backtrace.go#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `File.Close` returns an `error`. `stopper.Closer` does not. :(
</details>
ah. errcheck whitelist got me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7980)*
<!-- Sent from Reviewable.io -->
